singularity-eos/closure/mixed_cell_models.hpp

```diff
@@ -462,22 +462,87 @@ pte_closure_flag_with_line_impl(EOSIndexer &&eoss, const Real Volume, const Real
+PORTABLE_INLINE_FUNCTION static void pte_residual(const Real vfrac_tot, const Real utot,
+                                                  T1 &&vfrac, Real *u, T2 &&temp,
+                                                  T3 &&press, Real *residual) {
   Real vsum = 0.0;
   Real esum = 0.0;
   for (int m = 0; m < nmat; ++m) {
     vsum += vfrac[m];
     esum += u[m];
   }
-  residual[0] = 1.0 - vsum;
+  residual[0] = vfrac_tot - vsum;
   residual[1] = utot - esum;
   for (int m = 0; m < nmat - 1; ++m) {
     residual[2 + m] = press[m + 1] - press[m];
     residual[1 + nmat + m] = temp[m + 1] - temp[m];
   }
 }
 
+template <int nmat, typename EOSIndexer, typename RealIndexer, typename LambdaIndexer,
+          typename OFFSETTER>
+PORTABLE_INLINE_FUNCTION void
+try_ideal_pte(EOSIndexer &&eos, const Real vfrac_tot, const Real utot, const Real rho[nmat],
+              RealIndexer &&vfrac, RealIndexer &&sie, RealIndexer &&temp, RealIndexer &&press,
+              LambdaIndexer &&lambda, const OFFSETTER ofst, Real Cache[nmat][MAX_NUM_LAMBDAS]) {
+  Real A[nmat], B[nmat];
+  Real rhoBsum = 0.0;
+  Real Asum = 0.0;
+  for (int m = 0; m < nmat; m++) {
+    A[m] = vfrac[m] * press[m]/temp[m];
+    Asum += A[m];
+    B[m] = sie[m]/temp[m];
+    rhoBsum += rho[m] * B[m];
+  }
+
+  const Real Tequil = utot/rhoBsum;
+  const Real Pequil = Tequil * Asum/vfrac_tot;
+  const Real alpha = Pequil/Tequil;
+
+  Real etrial[nmat], vtrial[nmat], ttrial[nmat], ptrial[nmat];
+  for (int m = 0; m < nmat; m++) {
+    etrial[m] = B[m] * Tequil;
+    vtrial[m] = A[m] / alpha;
+  }
+  for (int m = 0; m < nmat; m++) {
+    ttrial[m] = eos[ofst(m)].TemperatureFromDensityInternalEnergy(rho[m]/vtrial[m], etrial[m], Cache[m]);
+    if (eos[ofst(m)].PreferredInput() ==
+        (thermalqs::density | thermalqs::specific_internal_energy)) {
+      ptrial[m] = eos[ofst(m)].PressureFromDensityInternalEnergy(rho[m]/vtrial[m], etrial[m], Cache[m]);
+    } else if (eos[ofst(m)].PreferredInput() ==
+               (thermalqs::density | thermalqs::temperature)) {
+      ptrial[m] = eos[ofst(m)].PressureFromDensityTemperature(rho[m]/vtrial[m], ttrial[m], Cache[m]);
+    }
+  }
+
+  // compare std dev of pressures and accept if they're reduced
+  // this is a bit hokey.  what's the right acceptance criteria?
+  Real sum_orig = 0.0;
+  Real sq_sum_orig = 0.0;
+  Real sum_new = 0.0;
+  Real sq_sum_new = 0.0;
+  for (int m = 0; m < nmat; m++) {
+    sum_orig += press[m];
+    sq_sum_orig += press[m] * press[m];
+    sum_new += ptrial[m];
+    sq_sum_new += ptrial[m] * ptrial[m];
+  }
+  const Real mean_orig = sum_orig / nmat;
+  const Real var_orig = sq_sum_orig / nmat - mean_orig * mean_orig;
+  const Real mean_new = sum_new / nmat;
+  const Real var_new = sq_sum_new / nmat - mean_new * mean_new;
+  if (var_new < var_orig) {
+    for (int m = 0; m < nmat; m++) {
+      press[m] = ptrial[m];
+      temp[m] = ttrial[m];
+      sie[m] = etrial[m];
+      vfrac[m] = vtrial[m];
+    }
+  } 
```

Comment:
I think it would be useful to standardize how the residual is calculated. I started a bit on this in my attempts to break out the PTE functionality so that the convergence iteration can be broken out for vector calls, but I didn't get very far. I don't think this needs to be done in this MR of course though, but I think it's something to think about for the future.

Reply:
updated this.  how does it look now?